File: get-calendar-id.js
#!/usr/bin/env node
/**
 * Script pour récupérer l'ID d'un calendrier Google
 * Utilise les credentials OAuth2 configurés
 */

const { google } = require('googleapis');
const readline = require('readline');

// Configuration OAuth2
const CLIENT_ID = process.env.GOOGLE_CALENDAR_CLIENT_ID;
const CLIENT_SECRET = process.env.GOOGLE_CALENDAR_CLIENT_SECRET;
const REDIRECT_URI = 'urn:ietf:wg:oauth:2.0:oob';

const SCOPES = ['https://www.googleapis.com/auth/calendar.readonly'];

if (!CLIENT_ID || !CLIENT_SECRET) {
  console.error('❌ Erreur: Variables d\'environnement manquantes');
  console.error('   Assurez-vous que GOOGLE_CALENDAR_CLIENT_ID et');
  console.error('   GOOGLE_CALENDAR_CLIENT_SECRET sont définis');
  console.error('\nChargez votre .env.local:');
  console.error('   source .env.local (Linux/Mac)');
  console.error('   ou créez les variables temporairement');
  process.exit(1);
}

const oauth2Client = new google.auth.OAuth2(
  CLIENT_ID,
  CLIENT_SECRET,
  REDIRECT_URI
);

function getAccessToken(oauth2Client, callback) {
  const authUrl = oauth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: SCOPES,
  });

  console.log('🔐 Autorisez cette application en visitant cette URL:\n');
  console.log(authUrl);
  console.log('\n');

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  rl.question('📋 Entrez le code de la page: ', (code) => {
    rl.close();
    oauth2Client.getToken(code, (err, token) => {
      if (err) return console.error('❌ Erreur lors de la récupération du token', err);
      oauth2Client.setCredentials(token);
      callback(oauth2Client);
    });
  });
}

async function listCalendars(auth) {
  const calendar = google.calendar({ version: 'v3', auth });

  try {
    console.log('\n📅 Liste de vos calendriers:\n');
    console.log('=' .repeat(80));

    const res = await calendar.calendarList.list();
    const calendars = res.data.items;

    if (!calendars || calendars.length === 0) {
      console.log('❌ Aucun calendrier trouvé.');
      return;
    }

    calendars.forEach((cal, index) => {
      console.log(`\n${index + 1}. 📆 ${cal.summary}`);
      console.log(`   ID: ${cal.id}`);
      console.log(`   Description: ${cal.description || 'Aucune'}`);
      console.log(`   TimeZone: ${cal.timeZone}`);
      console.log(`   Access: ${cal.accessRole}`);

      if (cal.primary) {
        console.log('   ⭐ CALENDRIER PRINCIPAL');
      }

      console.log('   -'.repeat(40));
    });

    console.log('\n' + '='.repeat(80));
    console.log('\n💡 Utilisez l\'ID du calendrier souhaité comme GOOGLE_CALENDAR_ID');
    console.log('   Pour le calendrier principal, vous pouvez utiliser: "primary"');
    console.log('   Pour un calendrier spécifique, copiez son ID (ex: xxx@group.calendar.google.com)\n');

    // Afficher un exemple de configuration
    const targetCalendar = calendars.find(cal => cal.summary === 'RDV Site') || calendars[0];
    console.log('📝 Exemple de configuration pour', targetCalendar.summary, ':\n');
    console.log('GOOGLE_CALENDAR_ID="' + targetCalendar.id + '"');
    console.log('');

  } catch (error) {
    console.error('❌ Erreur lors de la récupération des calendriers:', error.message);
  }
}

// Si on a déjà un refresh token, on l'utilise
if (process.env.GOOGLE_CALENDAR_REFRESH_TOKEN) {
  console.log('✅ Utilisation du refresh token existant\n');
  oauth2Client.setCredentials({
    refresh_token: process.env.GOOGLE_CALENDAR_REFRESH_TOKEN,
  });
  listCalendars(oauth2Client);
} else {
  console.log('🔑 Pas de refresh token trouvé, authentification nécessaire\n');
  getAccessToken(oauth2Client, listCalendars);
}

File: get-calendar-id.py
#!/usr/bin/env python3
"""
Script pour récupérer l'ID d'un calendrier Google
Utilise les credentials OAuth2 configurés
"""

import os
import sys
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from googleapiclient.discovery import build

# Scopes requis pour lire les calendriers
SCOPES = ['https://www.googleapis.com/auth/calendar.readonly']

def get_calendar_list():
    """Récupère et affiche la liste des calendriers"""

    # Récupérer les credentials depuis les variables d'environnement
    client_id = os.getenv('GOOGLE_CALENDAR_CLIENT_ID')
    client_secret = os.getenv('GOOGLE_CALENDAR_CLIENT_SECRET')
    refresh_token = os.getenv('GOOGLE_CALENDAR_REFRESH_TOKEN')

    if not client_id or not client_secret:
        print('❌ Erreur: Variables d\'environnement manquantes')
        print('   Assurez-vous que GOOGLE_CALENDAR_CLIENT_ID et')
        print('   GOOGLE_CALENDAR_CLIENT_SECRET sont définis')
        print('\nChargez votre .env.local:')
        print('   export $(cat .env.local | xargs)')
        sys.exit(1)

    creds = None

    # Si on a un refresh token, on l'utilise
    if refresh_token:
        print('✅ Utilisation du refresh token existant\n')
        creds = Credentials(
            token=None,
            refresh_token=refresh_token,
            token_uri='https://oauth2.googleapis.com/token',
            client_id=client_id,
            client_secret=client_secret,
            scopes=SCOPES
        )
    else:
        print('🔑 Pas de refresh token trouvé, authentification nécessaire\n')
        # Créer un fichier temporaire avec les credentials
        import json
        import tempfile

        credentials = {
            'installed': {
                'client_id': client_id,
                'client_secret': client_secret,
                'redirect_uris': ['urn:ietf:wg:oauth:2.0:oob'],
                'auth_uri': 'https://accounts.google.com/o/oauth2/auth',
                'token_uri': 'https://oauth2.googleapis.com/token'
            }
        }

        with tempfile.NamedTemporaryFile(mode='w', suffix='.json', delete=False) as f:
            json.dump(credentials, f)
            credentials_file = f.name

        try:
            flow = InstalledAppFlow.from_client_secrets_file(
                credentials_file, SCOPES)
            creds = flow.run_console()
        finally:
            os.unlink(credentials_file)

    # Construire le service Calendar
    try:
        service = build('calendar', 'v3', credentials=creds)

        # Récupérer la liste des calendriers
        print('📅 Liste de vos calendriers:\n')
        print('=' * 80)

        calendar_list = service.calendarList().list().execute()
        calendars = calendar_list.get('items', [])

        if not calendars:
            print('❌ Aucun calendrier trouvé.')
            return

        for index, calendar in enumerate(calendars, 1):
            print(f'\n{index}. 📆 {calendar.get("summary")}')
            print(f'   ID: {calendar["id"]}')
            print(f'   Description: {calendar.get("description", "Aucune")}')
            print(f'   TimeZone: {calendar.get("timeZone")}')
            print(f'   Access: {calendar.get("accessRole")}')

            if calendar.get('primary'):
                print('   ⭐ CALENDRIER PRINCIPAL')

            print('   ' + '-' * 40)

        print('\n' + '=' * 80)
        print('\n💡 Utilisez l\'ID du calendrier souhaité comme GOOGLE_CALENDAR_ID')
        print('   Pour le calendrier principal, vous pouvez utiliser: "primary"')
        print('   Pour un calendrier spécifique, copiez son ID (ex: xxx@group.calendar.google.com)\n')

        # Afficher un exemple de configuration
        target_calendar = next(
            (cal for cal in calendars if cal.get('summary') == 'RDV Site'),
            calendars[0]
        )

        print(f'📝 Exemple de configuration pour {target_calendar.get("summary")}:\n')
        print(f'GOOGLE_CALENDAR_ID="{target_calendar["id"]}"')
        print()

    except Exception as error:
        print(f'❌ Erreur lors de la récupération des calendriers: {error}')
        sys.exit(1)

if __name__ == '__main__':
    get_calendar_list()
